### ISSUE TITLE
Project status

### ISSUE THREAD

**Issue by rm-hull** [cloverage/cloverage#103]
Hi, @palfrey / @JacekLach

Seems like this project has been quite inactive lately and there are several open issues and PR's that would fix some minor problems.

If Lshift are no longer interested in maintaining this tool, perhaps ownership should be transferred to the community - possibly under a cloverage github organization where interested parties can provide active maintenance. What do you think about this?

cc: @mishadoff @FossiFoo @jarnovayrynen @cichli @ryfow @JustinC474 @lvh

**Comment by lvh** [comment]
I've wanted to write up something similar by finding a custodian first, but haven't gotten around to it, so thanks for writing this :)

It'd be pretty sad if it went to another location where it would once again become unmaintained. One of the avenues I've considered is talking to Cognitect -- mostly because this project would obviously benefit from having eyes on it that are also intimately familiar with the Clojure compiler, and, ideally also the ClojureScript one.

Regardless, somebody needs to start maintainering this. I can probably manage to spend some cycles on it, but not enough to actually maintain it full time. Perhaps with a handful of maintainers we can do a reasonable job of at least cleaning out the current PR/issue queue, but we'd (at least) need clojars push access and contributor access to this repo to do anything about it.

**Comment by FossiFoo** [comment]
I'd sure pitch in. This project is pretty relevant to my workflow and well-being. ;)

**Comment by JacekLach** [comment]
Indeed I don't really have the time to maintain the project anymore, since I haven't done anything in clojure in a long while.

@palfrey I believe the clojars login/password is in the travisci build settings, if you don't know what it is. I've made the https://github.com/cloverage team a while ago, but we never got to moving the project there; happy to hand it over to you if you're up for coordinating such a handover.

I'd be really happy to see someone pick this project up; it's a great feeling to find random projects that I use day-to-day have a cloverage dependency (puppetdb, for example :P).

**Comment by carrete** [comment]
@rfkm has kept up his own fork at https://github.com/rfkm/cloverage. S/he seems a likely candidate.

**Comment by lvh** [comment]
Cool! Fragmenting the cloverage namespace again seems unfortunate though... We haven't heard from anyone from lshift yet, but maybe pinging some people works: @david-lshift @ash-lshift @irene-lshift @IanRogers-LShift ? That seems important because they own the cloverage real estate...

@rfkm, would you be interested in moving your changes into one location?

If someone is willing to give me appropriate repo access, I'd be more than willing to point everything to cloverage/cloverage; then we can look at porting changes towards that repo.

**Comment by IanRogers-LShift** [comment]
Hi @lvh, @JacekLach  

LShift have had a bit of a chat in house (not much though) and Jacek's idea of moving it to https://github.com/cloverage to encourage a wider community seems like a good one. I'm not really involved with clojure at all any more but personally I'd prefer cloverage to thrive rather than be over-protective of it. @palfrey is kinda the guardian at the moment.

So if you do fork make sure to invite us :-)

Cheers

**Comment by palfrey** [comment]
As @IanRogers-LShift said :) My basic concern is who has ultimate authority for accepting PRs and doing new releases. There isn't really any internal LShift interest from anyone in particular in dealing with this at the moment, so if the community of folks that have been sending in PRs to this want to nominate someone to be in the driving seat, we'd probably be happy handing it over to them.

**Comment by lvh** [comment]
cloverage organization with a few maintainers (to reduce bus number) seems like a good idea; so I guess step one is to move the repo to cloverage/cloverage and make that the canonical repository.

I can probably spend some cycles this week at least clearing out the PR backlog.

**Comment by rfkm** [comment]
@lvh 
Sorry, I missed notifications.
Yeah, it sounds great :) Please feel free to move my changes to a new location.
I'm willing to help you do that if you need, while I also don't have much time to maintain the project.

**Comment by palfrey** [comment]
So, I've just spoke a bit further to folks here, and here's a proposal. I'm seeing a fair amount of enthusiasm here, but we're concerned that there might be the problem of people being initially enthusiastic but then dropping off pretty quickly. To try and combat that, and also so we've still got a little bit of control over this, here's what we're thinking:

* Project for the moment stays in the LShift organisation
* I add a couple of people as collaborators with "write" permissions (i.e. you can merge PRs). I'm thinking @lvh and possibly @rfkm if you feel like merging stuff back in?
* I still have the keys to do releases, but basically if the collaborators say "we've got a decent block of stuff ready, can you do a new release" I'm very likely to do so provided it doesn't look like a complete mess.
* We'd expect changes to have been reviewed by *someone* i.e. that people aren't making commits to master directly, but that everything involves Person A doing stuff, making a PR and Person B (and C, D, etc if wanted) reviewing. Different people at different times hopefully, but no code going in that's only had one person see it. I'm not going to get ruthless about that (and ironically may eventually directly add something to the README on master about all of this...), but it's a good practice that would make me much happier about pushing out new releases.

Does this work? We can consider moving this elsewhere eventually, but I'm guessing the priority is getting the project running again! If it's still working in a week or two, I'll probably add some text to the README about this (and a Code of Conduct, etc)

**Comment by lvh** [comment]
I was typing up something roughly equivalent, but first transferring it to the `cloverage` org, since the LShift folks have indicated that they're not interested in maintaining it on any timeline -- so the cloverage org seems more appropriate. Even if there is a marked drop-off in enthusiasm after e.g. cleaning the current backlog, having the repo live here doesn't fix the lack of maintainers. Similarly, it seems appropriate for release keys to not have a bus factor of 1, even if you end up cutting the actual releases for a while. Everything else, in particular re: reviews and CoC, sounds good though.

**Comment by bostonaholic** [comment]
I love that this conversation is happening. I'd love to see cloverage continue since I use it for several different OSS projects of mine. I'd also be interested in getting more involved in maintaining this tool.

I just created #cloverage in the Clojurians Slack channel to have more in-depth conversations.

**Comment by jarnovayrynen** [comment]
Happy to see this discussion taking place. I hope cloverage will get more active again so that it'll stay relevant. Maybe at first we should find a way clearing the current issue/PR queue. I can occasionally review some changes but I don't currently have time to become the actual maintainer.

**Comment by lvh** [comment]
Cool -- good to see we have ample volunteers :D I'm happy to pick up the suggested tasks, waiting on feedback from LShift :)

**Comment by palfrey** [comment]
Apologies for lack of responses over the past few days, as I've been dealing with one of our (paying) projects that needed all my time :smile: 

So, we're still keeping the project here for the time being. Part of this is a simple matter of publicity: we did a lot of work on this project, even if we're doing less so now, and so we'd like to still benefit from it being within our organisation. Another part of that is that we'd like to see how this goes - we're not currently focusing on this, in part because we're not currently working on a big Clojure project, but that might well change in the future. If there's a decent community who have tidied up things in a few months time, then this could be reconsidered, but for the moment we're keeping it here. Sorry for playing a certain amount of hardball on this one, but we don't want to massively reshuffle things until we see if this works well.

@lvh: I've just added you as a collaborator here. If you want to nominate others to also be collaborators, please shout and I'll do so.

**Comment by lvh** [comment]
@palfrey Thanks :) Does the current LShift control include not letting the new maintainer(s) cut releases? Can we agree on a concrete date and objective standard for meeting the community-being-formed criterion? I'd suggest 5 new reviewers, 10 PRs merged or somesuch by EOY. That seems relevant, because there's an obvious chicken-or-the-egg problem there with not-quite-maintainers don't at least have a path to true maintainerhood :)

For now several people have indicated that they are interested in reviewing, but not maintaining. @rm-hull @jarnovayrynen @bostonaholic @fossilfoo @rfkm -- if you review something, please don't hesitate to ping me loudly :) I'm currently in a plane over Uulanbaatar but I'll start cleaning up open PRs now.

**Comment by lvh** [comment]
@rfkm If you could submit individual features as PRs, that'd be very helpful.

**Comment by rfkm** [comment]
OK, I'll create PRs picking some changes of mine up.
I'm happy to see the revival of the project. Thank you, @lvh!

**Comment by lvh** [comment]
Thank _you_! I'm very excited to see these PRs!

**Comment by bostonaholic** [comment]
I start at CircleCI in two weeks. One of the things I might start looking into is how to get cloverage and circle to work together.

**Comment by lvh** [comment]
Awesome! Great collaboration between the both projects seems like a clear mutual benefit :)

**Comment by carrete** [comment]
> I start at CircleCI in two weeks. One of the things I might start looking into is how to get cloverage and circle to work together.

Yeah, awesome! We're Circle CI users and I'd love to see 1) junit xml output added, and 2) a non-zero exit status if coverage isn't at or above some threshold.

It's great to see a renewed interest in this project. Thanks everyone!

**Comment by bostonaholic** [comment]
> Yeah, awesome! We're Circle CI users and I'd love to see 1) junit xml output added, and 2) a non-zero exit status if coverage isn't at or above some threshold.

Those are exactly two things I'd like to see, too. I'm not making any pormises, but after I get settled in I'll start taking a look at these two features.

**Comment by palfrey** [comment]
@lvh: I don't currently have concrete criteria for that, no. OTOH, things look like they're doing very good so far! What is probably worth doing is having a conversation about this in say a month and reviewing how this is all doing, and then figuring out what to do about it then. This is all kinda being done on more of a gut feel about things than anything else TBH, and so I'd prefer to see how things settle out once we get past the initial honeymoon stage of everyone being very enthusiastic again ;)

**Comment by rm-hull** [comment]
I'm glad that this issue has revived interest in cloverage and has lead to this extended discussion. It is a tool which provides huge benefit to a lot of people, and it should not be left to bit-rot. 

I can't speak for anyone else, but personally, the worst thing that saps enthusiasm for contributing to open-source is spending time investigating & raising issues, fixing bugs and submitting PRs only for them to be completely ignored by the maintainers. This [comment](https://github.com/lshift/cloverage/issues/83#issuecomment-238882373) says it all really, and I'm paraphrasing here: _"no one replied, so we just moved on"_.

I am gracious that Lshift have given commit rights to someone else: this is important because there is less value in people maintaining their own divergent forks, and it _will_ help keep things alive.  However, I find comments along the line of _"there isn't really any internal LShift interest from anyone in particular in dealing with this"_ but then still wanting to maintain tight control a little bit disconcerting.

One surefire way to stall the enthusiasm and momentum that has built up so far is to block being able to cut releases; I can appreciate you have commercial pressures, but all the merged PRs + bug fixes add up to nought if we are still relying on someone with no motivation to push the button. 

Sorry if this all sounds harsh or negative, but I can't help but think that in order for this project to thrive, Lshift just needs to 'let go'.

**Comment by rm-hull** [comment]
Things I would like to see happen, in no particular order:
* zero-inbox policy on PRs and issues (I have no problem closing issues as 'wont fix' or rejecting PRs, but they should not be left to languish for months)
* Regular dependency updates
* Improve test coverage
* Support for Boot
* Refresh of the HTML reports L&F, using a templating engine like Hiccup
* Pluggable mechanism to support different backends - coveralls, circle-ci, raw, xml, whatever - so that it can be specified as a plugin dependency in the end-user project rather than baked into cloverage-core
* Move to a cloverage github organisation, to facilitate:
    - lein-cloverage and cloverage-core made first class projects in their own right
    - new boot-cloverage and plugin-XXX projects
* inclusive policy for bug-reporters, contributors & reviewers - if someone cares enough to raise an issue or submit a PR, then they should be trusted with commit-rights. 

It is still important to triage and code review and provide constructive criticism, but I'm very much on the side that to bring out the best in people you should trust them to do the right thing

**Comment by JacekLach** [comment]
> Refresh of the HTML reports L&F, using a templating engine like Hiccup

This is kinda offtopic, but relevant to mentioned idea of pulling in more dependencies. The way cloverage currently works is by injecting its own code (and dependencies) into the instrumented project. (Or alternatively pulling in the instrumented projects code into itself, whichever way is easier to reason about :P). Adding more dependencies to cloverage means more possible dependency clashes with instrumented projects, causing weird issues when the dependencies resolve differently when instrumented with cloverage than without (see for example #47).

The lazy approach to the problem (and what we've originally done) is try to keep the cloverage dependency tree very small.

The proper approach is to split into two libraries - a core or runtime, which gets injected into the instrumented project and is as small as at all possible, maybe even vendoring all of its dependencies so that they don't clash at all, handling only instrumentation and output of the 'raw' coverage data; and a richer wrapper (likely the lein plugin), which can have larger dependency trees and can handle everything else.

**Comment by carrete** [comment]
> I am gracious that Lshift have given commit rights to someone else: this is important because there is less value in people maintaining their own divergent forks, and it will help keep things alive. However, I find comments along the line of "there isn't really any internal LShift interest from anyone in particular in dealing with this" but then still wanting to maintain tight control a little bit disconcerting.

Same.

> It is still important to triage and code review and provide constructive criticism, but I'm very much on the side that to bring out the best in people you should trust them to do the right thing

Amen.

**Comment by lvh** [comment]
@palfrey Noted :) I can only echo the sentiments of other people above, but we'll revisit this in a month. One reason I would strongly prefer to have an objective criterion is because:

- the LShift PR reason you mentioned arguably doesn't change,
- one could easily argue that interest has died down when all that's happened is that we've burned through the initial backlog of changes that haven't gotten merged.

The part about cutting releases in my comment is more immediately relevant than the longer-term GitHub repository, though. Could you please add me to [the cloverage Clojars group](https://clojars.org/groups/cloverage) so I can at least cut new releases in the mean while? That, in combination with the revisiting mentioned above, seems like a reasonable compromise given that I'm now effectively the maintainer, and seems like it would assuage the concerns mentioned above.

**Comment by lvh** [comment]
@JacekLach That's an excellent point and I've tracked it in #109.

**Comment by lvh** [comment]
@rm-hull Excellent points, I have tried to make sure they're appropriately tracked, e.g. Boot support (#98), out-of-date dependencies (#110), HTML templating (#109 #111). Moving to new org is something I will follow up with as noted above.

**Comment by jarnovayrynen** [comment]
Good points, people. Especially @rm-hull, you have a solid argumentation and a plan as a way forward. And thanks @lvh for your good efforts maintaining this, keep it up!

On the feature side I'd like to see improved working with macros. In our project we are defining some vars with a macro when starting the application. We basically have a macro to read some values from Excel (user roles and their access rights coming from our customer) and define it as Clojure data. That stuff seems not to get defined at all during instrumentation currently. I'm not sure if our use case is a common one, but for us it kind of blocks using cloverage, because access right checks are all over the code base. Probably there are possibilities to work it around, but it hasn't been that high a priority for us yet.

**Comment by lvh** [comment]
@jarnovayrynen Makes sense. There are a few issues up around macros, including #85 and #82; they might be related to your issue. Do you have a minimal reproducing test case?

**Comment by FossiFoo** [comment]
i suggest tagging a release before actually refactoring a lot of the things i proposed. a proper version including codecov.io and lcov support would be great.

i guess boot support can go into a next version as well.

**Comment by lvh** [comment]
I agree completely, @FossiFoo. Alas, I can't cut releases; see the last 2 comments from @palfrey :) Concretely; I'd be happy to cut a release, but need to be added to the clojars group for it.

**Comment by FossiFoo** [comment]
Well, that's why i propose to do so here, so maybe @palfrey can go ahead if nobody objects.

**Comment by jarnovayrynen** [comment]
Sorry, been busy. As a matter of fact, I don't have a minimal test case
just yet. Basically it would be a file with a macro that defines vars. Then
parsing / instrumenting that file would fail. Let me see if I find time to
create one and I could create a branch.

**Comment by lvh** [comment]
Since several people have now asked me for releases over a period of 2 weeks to get their changes in a usable release; I've cut https://clojars.org/lvh/cloverage (lein artifact: `[lvh/cloverage "1.0.7"]`) pending my ability to cut real releases.

**Comment by carrete** [comment]
> I've cut https://clojars.org/lvh/cloverage (lein artifact: [lvh/cloverage "1.0.7"])

Thanks @lvh and everyone who submitted pull-requests.

**Comment by lvh** [comment]
I will also cut a `lvh/lein-cloverage` because `lein-cloverage` assumes `'cloverage` (aka `'cloverage/cloverage` is the One True Cloverage) -- right now `lein run -m cloverage.cloverage` works fine though :)

**Comment by rm-hull** [comment]
@lvh if you look at the deploy.sh script, this gets called from Travis when a commit is made... If it is on master it will do a deploy to clojars. 

If you change the version to 1.0.7 in both sub projects and the relevant dependencies on master and push, it _might_ just deploy without further ado.. 🤔

Also, tis worth amending the readme in the subdirs too as they are a bit stale.

**Comment by JacekLach** [comment]
The automatic build on master releases snapshot builds, but AFAIR can't release final versions because it can't sign them.
(though looking back at 1.0.6 it doesn't seem like it has been signed, and it looks like https://github.com/lshift/cloverage/blob/master/cloverage/project.clj#L14 disables signing releases since 7e2c9c49f1e846c2707eedc5545603ccc533b7e0, so maybe it will work)

**Comment by danielcompton** [comment]
Signing releases is no longer needed for Clojars, so this shouldn't be an issue.

**Comment by lvh** [comment]
@palfrey Friendly reminder; we really should cut a release.

**Comment by carrete** [comment]
> Friendly reminder; we really should cut a release.

Fork it and move on? As a point of reference, from someone who has used this tool in two day-jobs already as far back as two years ago, I consider this repository dead. We made the switch to @rfkm's fork back in December.

**Comment by FossiFoo** [comment]
Yes, it seems to me like a friendly fork to a cloverage organisition can not make this less maintainable.

I'm gone to Edgeland for the next 3 weeks, but i fully plan to go ahead with that -main rewrite afterwards.

**Comment by rm-hull** [comment]
@IanRogers-LShift @palfrey : its been nearly a month since you guys last commented on this thread; what's your take on things ? Are we ready to have that conversation and review how things are going ?

**Comment by carrete** [comment]
Is there some other place I can look to pick up the improvements that have been made recently?

**Comment by rm-hull** [comment]
Beginning to wonder if we should fork & rename; appealling to Lshift to cut a release has got precisely nowhere + they won't give the clojars creds to @lvh to be able to do that

**Comment by lvh** [comment]
Hi!

SNAPSHOTs have the current improvements. I'm currently taking time off (my mom is in town and I haven't seen her in a year), but I have a plan to resolve this when I get back, with real releases.


lvh

**Comment by palfrey** [comment]
Hi. So, as you might have all gathered, this hasn't exactly been our highest priority. The conclusion has therefore been finally reached to basically open this up properly and basically just hand it to the community. We'd like to be "backup admins" in the event of it all going quiet again, but in general will be leaving things alone for the foreseeable.

What does that mean concretely:
- @lvh - give me your Clojars login (lvh?) and I'll give you access to the Cloverage organisation there. Feel free to make releases as you please, add people as wanted, etc. Please leave the 'lshift' user in, but we promise not to make releases unless either a) everyone's gone quiet for multiple weeks and we have to pick up again or b) someone from LShift actually gets properly involved again, and then only with conversation with the wider community.
- I'm happy to transfer this repo to the Cloverage org (https://github.com/Cloverage). IIRC, I'll need to be added to that org to do that, and I'd quite like to get admin privs over there, again for the purposes of "if the currently active community goes entirely quiet", but I have no plans to do anything day-to-day and I leave that to the rest of you.

I think this is basically what was wanted overall by the community here. If there's other items, please say.

**Comment by lvh** [comment]
@palfrey Yep! `lvh` on cloverage. 

I'm very happy to hear about this change, because to be honest it was looking like a hard fork was the only way to solve this problem :)

**Comment by palfrey** [comment]
@lvh - I've added you to the cloverage and lein-cloverage groups on Clojars. I think that's everything we need to do at this point?

**Comment by lvh** [comment]
@palfrey Thanks! Can't think of anything. While obviously keeping the lshift account, I'm assuming it's OK if I do signed releases the usual way via my own Clojars account, that's fine?

**Comment by palfrey** [comment]
@lvh AFAIK, that should just work...

**Comment by lvh** [comment]
@palfrey Right; I'm expecting it to -- it's just that the current docs suggest that's not how you do it, and I'm trying to do the transition in a way that's mutually agreeable. I'll just cut the releases myself since that way I get to sign them.

**Comment by rm-hull** [comment]
Time for this thread to be closed, I think. 

Heres to the next chapter. Onwards and upwards 

:)
